server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveClusterAction.java

```diff
@@ -0,0 +1,455 @@
+            this.threadPool = threadPool;
+            this.clusterService = clusterService;
+            this.remoteClusterService = transportService.getRemoteClusterService();
+            this.indexNameExpressionResolver = indexNameExpressionResolver;
+            this.ccsCheckCompatibility = SearchService.CCS_VERSION_CHECK_SETTING.get(clusterService.getSettings());
+        }
+
+        @Override
+        protected void doExecute(Task task, ResolveClusterAction.Request request, ActionListener<ResolveClusterAction.Response> listener) {
+            if (ccsCheckCompatibility) {
+                checkCCSVersionCompatibility(request);
+            }
+            ClusterState clusterState = clusterService.state();
+            Map<String, OriginalIndices> remoteClusterIndices = remoteClusterService.groupIndices(
+                request.indicesOptions(),
+                request.indices()
+            );
+            OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+
+            Map<String, ResolveClusterInfo> clusterInfoMap = new ConcurrentHashMap<>();
+            if (remoteClusterIndices.size() > 0) {
+                int remoteRequests = remoteClusterIndices.size();
+                CountDown completionCounter = new CountDown(remoteRequests);
+                Runnable terminalHandler = () -> {
+                    if (completionCounter.countDown()) {
+                        listener.onResponse(new ResolveClusterAction.Response(clusterInfoMap));
+                    }
+                };
+
+                // add local cluster info if in scope of the index-expression from user
+                if (localIndices != null) {
+                    clusterInfoMap.put(
+                        RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY,
+                        new ResolveClusterInfo(true, false, hasMatchingIndices(localIndices, clusterState), Build.current())
+                    );
+                }
+
+                // make the cross-cluster calls
+                for (Map.Entry<String, OriginalIndices> remoteIndices : remoteClusterIndices.entrySet()) {
+                    String clusterAlias = remoteIndices.getKey();
+                    OriginalIndices originalIndices = remoteIndices.getValue();
+                    boolean skipUnavailable = remoteClusterService.isSkipUnavailable(clusterAlias);
+                    Client remoteClusterClient = remoteClusterService.getRemoteClusterClient(
+                        threadPool,
+                        clusterAlias,
+                        EsExecutors.DIRECT_EXECUTOR_SERVICE
+                    );
+                    ResolveClusterAction.Request remoteRequest = new ResolveClusterAction.Request(originalIndices.indices());
+                    remoteClusterClient.admin().indices().resolveCluster(remoteRequest, ActionListener.wrap(response -> {
```

Comment:
Do you really need `ActionListener#wrap` here? It doesn't look like the `onResponse` handler can throw, so I think you just want to create a `new ActionListener` directly. Also I'd suggest using `runAfter` (or `releaseAfter`) to do the decref.

Reply:
Fixed in [this commit](https://github.com/elastic/elasticsearch/pull/102726/commits/d1f6d81226d3416685abaae358a936369963675d).